pkg/streamingpromql/optimize/ast/prune_toggles.go

```diff
@@ -31,10 +47,34 @@ func (p *PruneToggles) Name() string {
 }
 
 func (p *PruneToggles) Apply(ctx context.Context, expr parser.Expr) (parser.Expr, error) {
-	return p.mapper.Map(ctx, expr)
+	p.mapper = NewPruneTogglesMapper()
+	p.pruneTogglesAttempts.Inc()
+	newExpr, err := p.mapper.Map(ctx, expr)
+	if p.mapper.HasChanged() {
+		p.pruneTogglesSuccesses.Inc()
+	}
+	return newExpr, err
+}
+
+func (p *PruneToggles) HasChanged() bool {
+	if p.mapper == nil {
+		return false
+	}
```

Comment:
When would `mapper` be `nil`?

Reply:
If HasChanged() is ever somehow called before Apply() (since we make the mapper there and not on initialisation of the struct), not really likely to happen but just a safeguard. Since we're not storing mapper in the struct anymore or keeping this method this doesn't matter now.